ratis-server-api/src/main/java/org/apache/ratis/server/raftlog/RaftLog.java

```diff
@@ -79,6 +80,10 @@ default ReferenceCountedObject<LogEntryProto> retainLog(long index) throws RaftL
   /**
    * @return null if the log entry is not found in this log;
    *         otherwise, return the {@link EntryWithData} corresponding to the given index.
+   *         The {@link EntryWithData} enclosed retained underlying resource. The client code need to ensure either
+   *         calling {@link ReferenceCountedObject#release()} on the result of
+   *         {@link EntryWithData#getEntryRef(TimeDuration)} or {@link EntryWithData#release()} if the entry
+   *         is discarded.
```

Comment:
Without updates, it won't call `getEntryRef` either, but `getEntry` which copies the returned `LogEntryProto`.
I think it's convenient that EntryWithData returns the ReferenceCountedObject itself via an interface because next step we will accept `StateMachine.read` return a ReferenceCountedObject. This means there will be logic to combine the refs of the LogEntry and the data somewhere, and it's convenient to put it inside a method of `EntryWithData`.

Reply:
> Without updates, the existing client code won't call getEntryRef either, but getEntry which copies the returned LogEntryProto?

I guest your mean below?
- `getEntryRef` -> `retainEntryWithData`;
- `getEntry` -> `getEntryWithData`.

Yes, the existing client code will keep calling `getEntryWithData` since the client code remains the same.

Reply:
> ...  it's convenient that EntryWithData returns the ReferenceCountedObject itself  ...

However, it is incompatible.

Also, if we add a new `release()` method to `EntryWithData`.  It becomes a different kind of `ReferenceCountedObject`. It becomes inconvenient (e.g. if we have some library code to release`ReferenceCountedObject`, it won't work for the new `EntryWithData` class.)